### PR TITLE
chore: Add metrics on types of filters in API

### PR DESCRIPTION
**Pull request by Artemka374**
## What ❔

Add metrics on types of filters, especially:
* Count of the current amount of filters of specific type
* Frequency of requests to specific filter type
* Lifetime of filters of specific type
* Request count for a specific filter

## Why ❔

We receive a lot of requests in our WebSocket API to getFilterChanges method.

We want to understand types of filters and distribution of load among them.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/metrics.rs:207]: It make sense to implement `From<&TypedFilter>` instead; no data from the filter is required to construct `FilterType`.

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/metrics.rs:218]: Bikeshedding: It doesn't look right to me to place this type in the `metrics` module. While all its additional fields *for now* are related to metrics, it isn't necessarily the case.

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/metrics.rs:221]: Bikeshedding: Please use more informative field names, such as `created_at` in this case.

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/metrics.rs:227]: Can't `guard` be constructed inside `new()` to make this method more robust?

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/metrics.rs:247]: Nit: You can add doc comments for metrics to make their purpose clearer; they are exported to Prometheus and can be displayed in Grafana tooltips etc.

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/metrics.rs:248]: For new metrics, you should specify `unit` if possible (e.g., `unit = Unit::Seconds` here and for `filter_lifetime`). The old metrics don't have `unit` set because this would change their names (by convention, unit is added as a suffix to the metric name, e.g. `api_filter_request_frequency_seconds`), so it would break existing Grafana dashboards.

File: core/lib/zksync_core/src/api_server/web3/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/mod.rs:70]: Is `pub` visibility required here? In general, it makes sense to restrict visibility to private, then to `pub(super)` / `pub(crate)`, and only if this doesn't work, make something public. Same goes for other types you define in this PR.

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:585]: Nit: You can use `?` to short-circuit this logic for readability:

```rust
let installed_filter = self.state.get(&index)?;
// ...
Some(&installed_filter.filter)
```

Also, it shouldn't compile without using `get_mut()` and changing the signature to something like `fn get_and_update(&mut self, index: U256)`. Alternatively (I'd prefer this option), you can move all mutation logic to `update()`.

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:607]: You shouldn't drop the previous value it'll lead to incorrect metrics, because semantically the filter is *updated*; it's still in place after the update. Instead, you should just update the `filter` field of the filter.

File: core/lib/zksync_core/src/api_server/web3/mod.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/mod.rs:18]: Bikeshedding: Although we don't enforce any particular `use` grouping for now, IMO, it makes sense to at least have a single `use` group for external crates in order to improve readability.

Comment (Artemka374) [reply core/lib/zksync_core/src/api_server/web3/mod.rs]: That was more because I haven't formatted the code properly

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:556]: This lint is usually resolved by naming a field with a `_` prefix, e.g. `_guard`.

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:570]: Why not use `Instant` type for this field as well?

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:580]: `self.request_count += 1` does the same thing and is more conventional.

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:590]: Nit: you can extract type into a local variable to not create it twice (the compiler may optimize it away, but it's easier to read in any case).

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:557]: An `Arc` is a code smell, necessary only because of the `Clone` implementation. You can remove `derive(Clone)` for `InstalledFilter` and `Filters` since it's actually unused (which is natural – `Filters` are enclosed into an `Arc<RwLock<_>>`).

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:636]: Nit: You don't seem to use the returned value anywhere, so it can be removed. Ditto for `update()`.

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:633]: I was thinking that it may be possible to pack updating stats and the filter together, but it's actually not the case (and even if it was possible, it probably makes sense to measure intervals between at the start of `getFilterChanges`, not at the end). So combining this logic with `get()` may make sense. A read lock on `Filters` can be acquire faster theoretically, but since in practice we never hold locks, the effect on performance should be minimal.

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/metrics.rs:217]: Bikeshedding: `api_web3_filter` looks like a more appropriate prefix; most existing Web3-related metrics have the `api_web3` prefix.

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (Artemka374) [reply core/lib/zksync_core/src/api_server/web3/metrics.rs:228]: That is true, I completely don't understand in which range it should be for now

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (Artemka374) [reply core/lib/zksync_core/src/api_server/web3/state.rs:557]: true, I was more focused on making new structures work and not changing old ones to not break anything

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (Artemka374) [reply core/lib/zksync_core/src/api_server/web3/state.rs:633]: Regarding that, do we need to have  `RwLock`  there at all in case we combine them?

Comment (Artemka374) [reply core/lib/zksync_core/src/api_server/web3/state.rs]: I'm thinking maybe we can just use `Mutex`, since we are not doing just read anywhere

File: core/lib/zksync_core/src/api_server/web3/metrics.rs
Comment (RomanBrodetski) [reply core/lib/zksync_core/src/api_server/web3/metrics.rs:228]: I'd do exponential until like 2^20

**Review by RomanBrodetski — COMMENTED** [review]
Looks good! Let's wait for @slowli for the final review

File: core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs:272]: Eh, isn't the previous behavior (the filter is **not** updated by the call) the correct one? Maybe, I wasn't clear in the previous comment here – I wanted you to add a comment why the filter shouldn't be updated, not to add logic for updating the filter.

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:32]: Nit: You can combine this `use` with the `use` clause below.

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:630]: Bikeshedding: IMO, inserting an empty line after each code line like here can harm readability. Of course, it's very hard (impossible?) to quantify the empty line rules, so feel free to ignore this comment. (IMO, an "empty line per ~5–10 statements" is a good ratio, but it's a personal preference.)

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (slowli) [inline core/lib/zksync_core/src/api_server/web3/state.rs:625]: Nit: It may make sense to clone `TypedFilter` right here (so, return `Option<TypedFilter>`); otherwise, a shared reference will borrow `self` mutably (this is determined by `&mut self`), which sort of defeats its purpose.

**Review by slowli — COMMENTED** [review]
Nice work! The only real question is changes in `get_filter_logs` (see the line comment).

File: core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs
Comment (Artemka374) [reply core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs:272]: Uh, yes, I guess did not understand you correctly

File: core/lib/zksync_core/src/api_server/web3/state.rs
Comment (Artemka374) [reply core/lib/zksync_core/src/api_server/web3/state.rs:630]: I am usually more preferring to divide logical blocks by empty lines or in places where it becomes hard to understand what is going on (in this place it is more about logical blocks)

**Review by slowli — APPROVED** [review]

**Pull request merged**
